### PR TITLE
WalletConnect

### LINKED ISSUE #499: implement WalletConnect (v1) as a dapp option into hdwallet
### Overview

Overview

https://walletconnect.com/

Wallet connect is a protocol allowing wallets to connect with Dapps. ShapeShift as a Dapp would like to allow any v1 integrated wallet in the wallet connect ecosystem to connect to app.shapeshift.com as a wallet, and utilize all supported features. These features are specific to ETH, and ETH compatible EVM’s. 

HDwallet is a collection of supported wallet modules that implement a common base class. This bounty is to create a module that supports wallet-connect as a wallet option and conform to this base class of HDwallet. 

Wallet connect provides the following example dapp
Example Dapp: https://github.com/WalletConnect/walletconnect-example-dapp

Example MetaMask addition PR
Example: https://github.com/shapeshift/hdwallet/pull/359/files

WalletConnect documentation:
Docs: https://docs.walletconnect.com/

### References and additional details

References

I am referencing the MetaMask integration and code snippets here to help explain the process of adding a new wallet to this HDwallet repo. As it is similar in scope/function. 

See it complete:
HDwallet sandbox:  https://hdwallet-shapeshift.vercel.app/
Screen Shot 2022-03-30 at 2 36 23 PM

Notice “Pair Wallet”

Wallet Connect will have a more interactive pairing process. 
Summary of pairing:

For testing manually I would recommend metamask mobile (supports wallet-connect) or trustwallet (mobile).

There will be a walletconnect modal displayed to the user to require the user to scan a QR code in their mobile wallet.

(there are also desktop wallet intents that can launch desktop apps automatically.) as well as intents to launch mobile wallets if the user is viewing the dapp on their mobile wallet.

Example dapp:https://example.walletconnect.org/

Screen Shot 2022-03-30 at 2 39 33 PM

After pairing, the sandbox should implement all supported functions, and display errors on those without support.

Note: asset support is flagged in the wallet class
https://github.com/shapeshift/hdwallet/blob/master/packages/hdwallet-metamask/src/metamask.ts#L23

Metamask package dir
https://github.com/shapeshift/hdwallet/tree/master/packages/hdwallet-metamask

Metamask Tests
https://github.com/shapeshift/hdwallet/blob/master/integration/src/wallets/metamask.ts

### Acceptance Criteria

A.C.
   * a complete wallet module with support for ETH functions
   * Some form of tests written into both unit/integration (with mocked response data)
   * A working sandbox wallet-connect option against a live wallet-connect server/wallet (tested manually)

### Need By Date

no

### Screenshots/Mockups

N/A

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Estimated effort

2-3 days

### Sponsor / Stakeholder

@bithighlander

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by pastaghost**
##  Description
This adds a package for integrating WalletConnect as an adapter. This provides out-of-the box support for several mobile and desktop wallets not currently natively supported on app.shapeshift.com.

## Issue
Closes #499

### CODE REVIEW

**Review by gomesalexandre — COMMENTED** [review]
This LGTM overall. As a full-featured hdwallet, it has some lower-level HD primitives which I'm not familiar with so stamping this will come down to testing. 
How can it be tested, should we run the Sandbox or is any web integration needed?

File: packages/hdwallet-walletconnect/package.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  "source": "src/index.ts",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "build": "tsc --build",
+    "clean": "rm -rf dist tsconfig.tsbuildinfo",
+    "dev": "yarn tsc --build --watch",
+    "prepublishOnly": "yarn clean && yarn build"
+  },
+  "dependencies": {
+    "@shapeshiftoss/hdwallet-core": "1.24.1-alpha.0",
```
Comment (0xdef1cafe) [inline packages/hdwallet-walletconnect/package.json:18]: Update?

**Review by 0xdef1cafe — DISMISSED** [review]

**Review by 0xdef1cafe — APPROVED** [review]

File: packages/hdwallet-walletconnect/src/walletconnect.ts [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+  readonly _supportsETHInfo = true;
+  readonly _supportsBTCInfo = false;
+  readonly _supportsBTC = false;
+  readonly _isWalletConnect = true;
+  readonly _supportsEthSwitchChain = false;
+
+  info: WalletConnectWalletInfo & core.HDWalletInfo;
+  provider: WalletConnectProvider;
+  connected = false;
+  chainId = -1;
```
Comment (gomesalexandre) [inline packages/hdwallet-walletconnect/src/walletconnect.ts:120]: Comment this?

File: packages/hdwallet-walletconnect/src/walletconnect.ts [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+  }
+
+  public async initialize(): Promise<void> {
+    /** Subscribe to EIP-1193 events */
+    this.provider.connector.on("session_update", async (error, payload) => {
+      if (error) {
+        throw error;
+      }
+
+      const { chainId, accounts } = payload.params[0];
```
Comment (gomesalexandre) [inline packages/hdwallet-walletconnect/src/walletconnect.ts:156]: Destructure `payload.params[0]` as a named const?

**Review by gomesalexandre — APPROVED** [review]
Smallish comments but overall LGTM.

Untested in absence of testing steps, but this will be tested in web when installing it there.

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**